Firestore/core/src/local/leveldb_target_cache.cc

```diff
@@ -269,6 +292,24 @@ void LevelDbTargetCache::RemoveAllKeysForTarget(TargetId target_id) {
   }
 }
 
+void LevelDbTargetCache::RemoveQueryTargetKeyForTargets(
+    const std::unordered_set<TargetId>& target_ids) {
+  std::string index_prefix = LevelDbQueryTargetKey::KeyPrefix();
+  auto index_iterator = db_->current_transaction()->NewIterator();
+  index_iterator->Seek(index_prefix);
+
+  LevelDbQueryTargetKey row_key;
+  for (; index_iterator->Valid(); index_iterator->Next()) {
```

Comment:
This is pessimizing target deletion considerably, making it O(targets) where it was previously O(targets-to-delete). Previously, `RemoveTarget` would construct the query-target key from the target data but now we're scanning all query-target rows.

It seems like we could try to construct the query-target key to delete based on the proto data and do like RemoveTarget always did.

Indeed I'm surprised that this this PR didn't just add a `RemoveTarget` overload that operated on a const message reference instead of changing this around to reimplement deletion.

Ultimately the workaround we're adding here is for a problem where the query proto has some invalid data in there. We haven't proven that this is happening because of general data corruption issues so I'm disinclined to rewrite all of this code in a way that doesn't trust the relationships built into the data. We should just not validate the Query part of the TargetData when removing a target.

Reply:
I agree with your concerns, but I went with this approach for two reasons:

- We don't actually store the Canonical ID. Instead, we compute it from the Target model. That means that we have to fully deserialize into the model. If we don't validate, we end up with a Model that has an invalid field ordering, which we would then use to create a Canonical ID that is different from the original ID (unless they happen to both be corrupt). Hence, we don't know whether we would be able to delete the row from Level DB.

- The LRU runs is already O(n) as it has to read all targets to collect the sequence numbers.

That's all I got for now :)

Reply:
Somehow I thought we stored the canonical ID, but you're right: the only place it exists is in the keys of the query-target mapping.

Your point re O(n) is valid, but it's a different "table" so it's more I/O. If there were a way to avoid this I'd push for it, but I can see how you arrived here.

I'm still concerned that regular deletion and GC deletion follow different code paths but it's definitely not worth pessimizing single target removal to make this line up, nor is it worth writing a migration to force these to line up. I guess this is as good as it's going to get.

Reply:
Regular deletion is actually not used for LevelDB, so we could use the new code path for `RemoveTarget` without much harm. Let me know if you'd like me to do so.